src/systems/filecoin_blockchain/storage_power_consensus/_index.md

```diff
@@ -44,70 +46,162 @@ More precisely,
+        } else {
+            rand = seeds[len(seeds)-1].Randomness
+        }
+    }
+    return rand
+}
+
+GetBeaconEntriesForEpoch(epoch) []BeaconEntry {
+
+    // special case genesis: the genesis block is pre-generated and so cannot include a beacon entry 
+    // (since it will not have been generated). Hence, we only start checking beacon entries at the first block after genesis.
+    // If that block includes a wrong beacon entry, we simply assume that a majority of honest miners at network birth will
+    // simply fork.
+    entries := []
+    if epoch == 0 {
+        return entries
+    }
+
+    maxDrandRound := MaxBeaconRoundForEpoch(epoch)
+
+    // if checking the first post-genesis block, simply fetch the latest entry.
+    if epoch == 1 {
+        rand := drand.Public(maxDrandRound)
+        return append(entries, rand)
+    }
+
+    // for the rest, fetch all drand entries generated between this epoch and last
+    prevMaxDrandRound := MaxBeaconRoundForEpoch(epoch - 1)
+    if (maxDrandRound == prevMaxDrandRound) {
+        // no new beacon randomness
+        return entries
+    }
+
+    entries := []
+    curr := maxDrandRound
+    for curr > prevMaxDrandRound {
+        rand := drand.Public(curr)
+        entries = append(entries, rand)
+        curr -= 1
+    }
+    // return entries in increasing order
+    reverse(entries)
+    return entries
+}
+```
 
-- The {{<sref sector_sealer>}} uses tickets as SealSeeds to bind sector commitments to a given subchain.
-- The {{<sref post_generator>}} likewise uses tickets as PoStChallenges to prove sectors remain committed as of a given block.
-- They are drawn by the Storage Power subsystem as randomness in {{<sref leader_election>}} to determine their eligibility to mine a block
-- They are drawn by the Storage Power subsystem in order to generate new tickets for future use.
+{{<label beacon_entry_validation>}}
+### Validating Beacon Entries on block reception
+
+Per the above, a Filecoin chain will contain the entirety of the beacon's output from the Filecoin genesis to the current block.
+
+Given their role in leader election and other critical protocols in Filecoin, a block's beacon entries must be validated for every block. See {{<sref drand>}} for details. This can be done by ensuring every beacon entry is a valid signature over the prior one in the chain, using drand's [`Verify`](https://github.com/drand/drand/blob/master/beacon/beacon.go#L72) endpoint as follows:
+
+```go
+// This need not be done for the genesis block
+// We assume that blockHeader and priorBlockHeader are two valid subsequent headers where block was mined atop priorBlock
+ValidateBeaconEntries(blockHeader, priorBlockHeader) error {
+    currEntries := blockHeader.BeaconEntries
+    prevEntries := priorBlockHeader.BeaconEntries
+
+    maxRoundForEntry := MaxBeaconRoundForEpoch(blockHeader.Epoch)
+    // ensure entries are not repeated in blocks
+    lastBlocksLastEntry := prevEntries[len(prevEntries)-1]
```

Comment:
What happens currently in Lotus is: 
* no drand value / information in the genesis block
* first epoch block contains one drand entry which is _not_ verifiable by itself since the entry does not contain the previous signature. We can't put a entry in the genesis block because we dont know it in advance (while the genesis block is quite predetermined I guess)
* At the second epoch, the "main chain" 2-nd epoch block contains a drand entry whose previous signature is the drand signature put in the epoch 1 block.

**consequence**: Miners will build on the first block without knowing if the randomness is valid or not. However, on the second block, miners will be able to verify if the randomness on the first round is correct. If not, that's a "re-org" as usual.

So here if `priorBlockHeader.Height == 0` we must skip that verification and return nil directly. This is what Lotus is doing [here](https://github.com/filecoin-project/lotus/blob/master/chain/beacon/drand/drand.go#L169).

That being said, there are other potential ways to handle genesis.  For example, in the epoch 1 header, include the full beacon entry with the previous signature as well.

Reply:
right, I didn't carry through my change appropriately from `GetBeaconEntriesForEpoch`. Thanks. Should be fixed now.